cpp/src/parquet/encoding.h

```diff
@@ -296,13 +296,16 @@ class TypedDecoder : virtual public Decoder {
 
     // Depending on the number of nulls, some of the value slots in buffer may
     // be uninitialized, and this will cause valgrind warnings / potentially UB
+    // TODO: With SIMD shuffle in the loop below we might be able to get better
+    // performance on doing the memset only on values that aren't copied, and fill
+    // in all the values below.
     memset(static_cast<void*>(buffer + values_read), 0,
            (num_values - values_read) * sizeof(T));
 
     // Add spacing for null entries. As we have filled the buffer from the front,
     // we need to add the spacing from the back.
     int values_to_move = values_read;
-    for (int i = num_values - 1; i >= 0; i--) {
+    for (int i = num_values - 1; i > (values_to_move - 1) && values_to_move > 0; i--) {
       if (BitUtil::GetBit(valid_bits, valid_bits_offset + i)) {
```

Comment:
Perhaps a `ReverseBitmapReader` class would make this faster? Or perhaps not...

Reply:
I'd rather keep this limited.  I'm also not sure where the gains would materialize from it?

Reply:
It may be slightly faster than `GetBit` calls. Though I'm not sure.